### PR TITLE
[Micro:Bit] Define valid and reserved pins for Micro:Bit

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR defines valid, reserved, and invalid pins for the Micro:Bit (MB) which are different from the Circuit Playground (CP) Express and Classic.

On the MB, valid pins are [0, 1, 2] and there are no reserved pins. On the other hand:
Circuit Playground Express -  valid pins: ['A0', 'A1', 'A4', 'A5', 'A6'], reserved pins: ['A2', 'A3', 'A7']
Circuit Playground Classic - valid pins: [0, 2, 3, 6, 12], reserved pins: [1, 9, 10].
[This PR](https://github.com/code-dot-org/code-dot-org/pull/47052) defined the reserved pins, and [this PR](https://github.com/code-dot-org/code-dot-org/pull/47243) defined the valid pins for the Circuit Playground.

In `maker/commands.js` I defined a function called `isMicroBitBoard` that uses the `board.boardClient_` value to determine if the board is a MB (returns `true`) or CP (returns `false`). This boolean is then passed as a parameter in `apiValidateType`. 

Before this update:

https://user-images.githubusercontent.com/107423305/219538484-dbf17257-6382-4c03-b972-c9e605950b95.mp4

After this update:

https://user-images.githubusercontent.com/107423305/219538790-4af3fe6f-e7b3-48d2-a0bd-c56d7a6dee90.mp4

And here's a video of the program running with pin 1 connected with an external color LED:

https://user-images.githubusercontent.com/107423305/219542248-66284c04-f6b3-4d77-bf17-01b67c5f2812.mp4

## Links

- jira ticket: [[Micro:bit] Fix pin mappings for Micro:bit](https://codedotorg.atlassian.net/browse/SL-558)

## Testing story
I added a couple unit tests in `commandsTest.js` to test that pin 1 is a valid pin on a M:B and pin 3 is invalid on a M:B. 
I chose to use the `MicrobitStubBoard` for these unit tests instead of `FakeBoard` after discussion with @epeach because `FakeBoard` was initially designed to be board-agnostic but is essentially a 'fake' Circuit Playground. Follow-up work is an investigation and possible refactoring of `FakeBoard`. 

## Follow-up work
As mentioned above in 'Testing story', follow-up work is to investigate `FakeBoard` which was initially created so that a user can run a program when the maker toolkit is enabled even when a board is not connected. `FakeBoard` is also being used in unit tests and now that we are supporting the M:B, we can investigate if we should continue using `FakeBoard` and perhaps refactor into `FakeCircuitPlaygroundBoard` and `FakeMicroBitBoard` or use the existing stub boards.
 

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/lib/kits/maker/commands.js [hunk 1]
```diff
@@ -6,12 +6,12 @@ import {
 import {BOARD_EVENT_ALIASES} from './boards/circuitPlayground/PlaygroundConstants';
 import MicroBitBoard from './boards/microBit/MicroBitBoard';
 
-/** @private {CircuitPlaygroundBoard} */
+/** @private {CircuitPlaygroundBoard, MicroBitBoard} */
```
Comment (fisher-alice) [inline apps/src/lib/kits/maker/commands.js]: I would like to confirm lines 9 and 14 contain the correct syntax for this private variable.

File: apps/src/lib/util/javascriptMode.js [hunk 2]
```diff
@@ -42,7 +42,8 @@ export function apiValidateType(
   varName,
   varValue,
   expectedType,
-  opt
+  opt,
```
Comment (sanchitmalhotra126) [inline apps/src/lib/util/javascriptMode.js:45]: This isn't part of your change, but do you know what the difference between `opts` and `opt` is?

File: apps/src/lib/util/javascriptMode.js [hunk 3]
```diff
@@ -42,7 +42,8 @@ export function apiValidateType(
...
+  board
```
Comment (sanchitmalhotra126) [inline apps/src/lib/util/javascriptMode.js]: Would it be better to pass an `isMicroBit` flag or maybe `boardType` so you don't have to pass in the whole `board` object?

File: apps/test/unit/lib/kits/maker/boards/makeStubBoard.js [hunk 4]
```diff
@@ -17,12 +17,23 @@ export class MicrobitStubBoard {
     this.updateListeners = [];
     this.analogChannel = new Array(16).fill(0);
     this.myPort = null;
+    this.boardClient_ = null;
+  }
+
+  boardClient_() {
+    this.boardClient_ = {};
```
Comment (sanchitmalhotra126) [inline apps/test/unit/lib/kits/maker/boards/makeStubBoard.js:24]: Curious, why do we need this function? Can we assign `this.boardClient = {}` in the constructor?

File: apps/src/lib/util/javascriptMode.js (hunk 2 above)
Comment (fisher-alice) [reply apps/src/lib/util/javascriptMode.js:45]: `opts` is an object and `opt` is a boolean and the value for `true` meaning 'Optional`.

File: apps/src/lib/util/javascriptMode.js (hunk 3 above)
Comment (fisher-alice) [reply apps/src/lib/util/javascriptMode.js]: I actually first wrote this by adding a local constant within each called `isMicroBitBoard` but had to do it multiple times. I'll create a function called `isMicroBitBoard` instead. Thanks!

File: apps/test/unit/lib/kits/maker/boards/makeStubBoard.js (hunk 4 above)
Comment (fisher-alice) [reply apps/test/unit/lib/kits/maker/boards/makeStubBoard.js:24]: I tried this at first, but the tests failed. When I printed the `stubBoardController` after assigning `this.boardClient_ = {}` in the constructor, this was the output:
```
LOG: MicrobitStubBoard{connect: connect, pinMode: pinMode, setPinMode: setPinMode, digitalWrite: digitalWrite, digitalRead: digitalRead, analogWrite: analogWrite, analogRead: analogRead, displayShow: displayShow, displayPlot: displayPlot, displayClear: displayClear, scrollString: scrollString, scrollInteger: scrollInteger, receivedEvent: receivedEvent, receivedAnalogUpdate: receivedAnalogUpdate, trackDigitalPin: trackDigitalPin, addFirmataEventListener: addFirmataEventListener, addFirmataUpdateListener: addFirmataUpdateListener, streamAnalogChannel: streamAnalogChannel, stopStreamingAnalogChannel: stopStreamingAnalogChannel, setDigitalOutput: setDigitalOutput, getBoardClientWithoutPort: getBoardClientWithoutPort, disconnect: disconnect, reset: reset, trackDigitalComponent: trackDigitalComponent, enableLightSensor: enableLightSensor, connectBoard: connectBoard, clearChannelData: clearChannelData}
```
There is no `boardClient_` key. So this is why I added the function. Do you have any ideas why this is so?

File: apps/src/lib/kits/maker/commands.js (hunk 1 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/commands.js]: @sanchitmalhotra126 lmk what you think.

File: apps/src/lib/util/javascriptMode.js [hunk 5]
```diff
@@ -64,25 +65,32 @@ export function apiValidateType(
...
-          1,
-          2,
-          3,
-          6,
-          9,
-          10,
-          12
-        ];
-        var reservedPins = ['A2', 'A3', 'A7', 1, 9, 10];
+        var validPins, reservedPins;
```
Comment (sanchitmalhotra126) [inline apps/src/lib/util/javascriptMode.js]: nit: you could default assign these to either the CircuitPlayground/Microbit values and then override them if needed, i.e.
```
var validPins = [A0, A1, ...]
var reservedPins = [A2, A3 ...]
if (isMicroBitBoard) {
   validPins = [0, 1, 2] ...
}
```

File: apps/src/lib/kits/maker/commands.js (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/lib/kits/maker/commands.js]: Hmm good question, I don't know if we have a good standard for union types like this in plain JavaScript...I think `{CircuitPlaygroundBoard | MicroBitBoard}` (borrowing [TypeScript union type syntax](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types)) would make the most sense to me at least?

File: apps/test/unit/lib/kits/maker/boards/makeStubBoard.js (hunk 4 above)
Comment (sanchitmalhotra126) [reply apps/test/unit/lib/kits/maker/boards/makeStubBoard.js:24]: Hmm, interesting...I'm not sure why it wouldn't be defined if not in the constructor. I'm okay to leave it as is for now then and we can take a look offline

**Review by sanchitmalhotra126 — APPROVED** [review]
Couple of small comments but LGTM otherwise!

File: apps/src/lib/kits/maker/commands.js (hunk 1 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/commands.js]: Thank you! This makes sense. Will use.

File: apps/src/lib/util/javascriptMode.js (hunk 5 above)
Comment (fisher-alice) [reply apps/src/lib/util/javascriptMode.js]: Sure - will do!

**Pull request merged**